vector/test.c

```diff
@@ -0,0 +1,81 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "array.c"
+#include "test.h"
+
+void test_vector_create(){
+  array* arr =  vector_create();
+  ASSERT(arr != NULL);
+  ASSERT(arr->capacity == init_capacity);
+  ASSERT(arr->size == 0);
+
+  vector_free(arr);
+}
```

Comment:
How can you know that this will pass - `ASSERT(arr != NULL);`?

Actually... you can't. Imagine a situation you are working on a big project, and there are tons of tests. Whenever you push to the remote git repository, the test build occurs.

Such "test build" launches the tests one after another. But... the machine that launches those tests, might also simultaneously launch other builds - e.g. tests for commits made by other people working in your team.

There might be a situation on which the machine would simply run OOM (out of memory) - e.g. if there is no limit for simultaneous builds. **In such situation, this test would fail on this assertion.**

And... it is hard to make this "work in all cases". You could for example **mock** (read [here](https://en.wikipedia.org/wiki/Mock_object) or [here](http://stackoverflow.com/questions/2665812/what-is-mocking)) the `malloc`, `free` and `realloc` (and eventually `calloc`) calls, so that they would work differently. 

This is doable e.g. thanks to `LD_PRELOAD` environment variable (read [here](http://stackoverflow.com/questions/426230/what-is-the-ld-preload-trick) or for real explanation - [here](http://man7.org/linux/man-pages/man8/ld.so.8.html) - under `LD_PRELOAD`).

I won't ask you to do that, because it is a total overkill for this task and I guess noone does it (because another hard question would have to be answered - how to mock those calls properly, so the tests would work always?).